### PR TITLE
support for headless services and pod annotations.

### PR DESCRIPTION
**Pull request by guptaarvindk**
Signed-off-by: Arvind Gupta <guptaarvindk@gmail.com>

**What this PR does / why we need it**:
support for:
headless services for master and slave.
annotations in master slave statefulsets.

in some appliance/kubernetes distributions few parameters are passed via annotations. So options to add annotation help.

Headless services are supported by many kubernetes distributions, which makes life easy within the cluster.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: stable/mariadb/README.md [hunk 1]
```diff
@@ -53,6 +53,7 @@ The following table lists the configurable parameters of the MariaDB chart and t
 | `image.pullPolicy`                        | MariaDB image pull policy                           | `Always` if `imageTag` is `latest`, else `IfNotPresent`           |
 | `image.pullSecrets`                       | Specify image pull secrets                          | `nil` (does not add image pull secrets to deployed pods)          |
 | `service.type`                            | Kubernetes service type                             | `ClusterIP`                                                       |
+| `service.clusterIp`                       | Specific cluster IP when service type is cluster IP. Use None for headless service |                                                                   |
```
Comment (juan131) [inline stable/mariadb/README.md]: You need to indicate the default value or `nil` if the property is not set by default.

File: stable/mariadb/README.md [hunk 2]
```diff
@@ -62,6 +63,8 @@ The following table lists the configurable parameters of the MariaDB chart and t
 | `replication.enabled`                     | MariaDB replication enabled                         | `true`                                                             |
 | `replication.user`                        | MariaDB replication user                            | `replicator`                                                       |
 | `replication.password`                    | MariaDB replication user password                   | _random 10 character alphanumeric string_                         |
+| `master.annotations[].key`                | key for the the annotation list item                |                                                    |
```
Comment (juan131) [inline stable/mariadb/README.md]: Could you use `master.podAnnotations` as an object? So the use can specify additional annotations. Check how it's developed on the [Redis chart](https://github.com/helm/charts/tree/master/stable/redis)

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for this contribution. Please take a look to my comments.

File: stable/mariadb/README.md (hunk 1 above)
Comment (guptaarvindk) [reply stable/mariadb/README.md]: Thanks juan131 , will do that.

File: stable/mariadb/README.md (hunk 2 above)
Comment (guptaarvindk) [reply stable/mariadb/README.md]: Thanks For the review juan131. `master.podAnnotations` in redis chart is surely simple way, but as I understand it takes multiple annotations as text string. While the way I am using it, it allows an array of annotations. To me array is more convenient what to specify, for example:

```
master:
  annotations:
  - key: endpoint0
    value: network0
  - key: endpoint1
    value: network1
```

Although, I am not too picky about this, and if redis way looks better to you we can do that way as well.

Thanks,
-Arvind

Comment (juan131) [reply stable/mariadb/README.md]: Could you share an example of use? How will you indicate your annotations when running the command below?

```
helm install --name my-release stable/mariadb --set XXX=xxx, YYY=yyy, ZZZ=zzz 
```

Comment (guptaarvindk) [reply stable/mariadb/README.md]: Hi @juan131 , sorry for late response. I got dragged in to other work. 

when using multiple array like I mentioned above, its better and easier to create cusom-values.yaml.  But still above mentioned annotations can be passed using --set as follows:

```
helm install --set master.annotations[0].key=endpoint0,master.annotations[0].value=network0,master.annotations[1].key=endpoint1,master.annotations[1].value=network1  stable/mariadb
```
output will be:
```
  template:
    metadata:
      annotations:
        endpoint0: 'network0'
        endpoint1: 'network1'
```
Does that solves your concern?

Thanks,

**Comment by carrodher** [comment]
Before continue, can you take a look at the errors in the test?
https://circleci.com/gh/helm/charts/22998?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by guptaarvindk** [comment]
Hi @carrodher ,

Sorry I am new to checkins. But how do I update the helm chart version? I mean there are multiple people checking in if everyone is bumping up the version on there own , that means a chess. I was assuming helm chart version is auto upgraded by the CI. Can you please point me whats the best way to handle it?
Thanks,

**Comment by carrodher** [comment]
Of course, I hope the following explanation resolves your doubts. You should edit the Chart.yaml file (`charts/stable/mariadb/Chart.yaml`) increasing the version according to [semver](http://semver.org/):

> Given a version number `MAJOR.MINOR.PATCH`, increment the:
> - MAJOR version when you make incompatible API changes,
> - MINOR version when you add functionality in a backward-compatible manner, and
> - PATCH version when you make backward-compatible bug fixes.
> Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

For example
```diff
name: mariadb
- version: 4.4.2
+ version: 4.4.3
appVersion: 10.1.36
description: Fast, reliable, scalable, and easy to use open-source relational database system. MariaDB Server is intended for mission-critical, heavy-load production systems as well as for embedding into mass-deployed software. Highly available MariaDB cluster.
```

You should include this change even if you add a comma to the readme. This process is not automated because the developer who creates the PR must determine if the change needs a MAJOR, MINOR or PATCH version.

**Comment by guptaarvindk** [comment]
Hi @carrodher being new to public GitHub contribution flow, when rebasing my changes form the upstream master I and up checking in all those differences to this branch in my last commit. Now I have fixed my branch to reflect only my commits. Can you please take a look if it looks correct now?

**Comment by guptaarvindk** [comment]
Hi @juan131 can you please review the latest pull request.

File: stable/mariadb/README.md (hunk 1 above)
Comment (juan131) [reply stable/mariadb/README.md]: It is still pending, isn't it?

File: stable/mariadb/README.md (hunk 2 above)
Comment (juan131) [reply stable/mariadb/README.md]: Yes thanks!!! Please add the default value on the third column of the table.

File: stable/mariadb/values-production.yaml [hunk 3]
```diff
@@ -75,6 +76,11 @@ replication:
   forcePassword: true
 
 master:
+  ## annotations
```
Comment (juan131) [inline stable/mariadb/values-production.yaml]: Use sth like:

```
  ## Mariadb Master additional pod annotations
  ## ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/
  # annotations:
  # - key: key1
  #    value: value1
```

**Review by juan131 — COMMENTED** [review]
Can you take a look to my suggestions? Thanks for the contribution

File: stable/mariadb/README.md (hunk 1 above)
Comment (guptaarvindk) [reply stable/mariadb/README.md]: yes thanks, Sorry I missed that.

**Comment by guptaarvindk** [comment]
Hi @juan131, Thanks for reviewing.  I fixed all those change. Can you please take a look again.

**Review by juan131 — APPROVED** [review]
Hi @guptaarvindk 

Thanks so much it!

**Pull request merged**